pkg/scheduler/admission/podnodeconstraints/admission.go

```diff
@@ -0,0 +1,77 @@
+package podnodeconstraints
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"reflect"
+
+	configlatest "github.com/openshift/origin/pkg/cmd/server/api/latest"
+	"github.com/openshift/origin/pkg/scheduler/admission/podnodeconstraints/api"
+
+	admission "k8s.io/kubernetes/pkg/admission"
+	kapi "k8s.io/kubernetes/pkg/api"
+	kclient "k8s.io/kubernetes/pkg/client/unversioned"
+)
+
+func init() {
+	admission.RegisterPlugin("PodNodeConstraints", func(c kclient.Interface, config io.Reader) (admission.Interface, error) {
+		pluginConfig, err := readConfig(config)
+		if err != nil {
+			return nil, err
+		}
+		return NewPodNodeConstraints(pluginConfig), nil
+	})
+}
+
+func NewPodNodeConstraints(config *api.PodNodeConstraintsConfig) admission.Interface {
+	return &podNodeConstraints{
+		config:  config,
+		Handler: admission.NewHandler(admission.Create),
+	}
+}
+
+type podNodeConstraints struct {
+	*admission.Handler
+	config *api.PodNodeConstraintsConfig
+}
+
+func readConfig(reader io.Reader) (*api.PodNodeConstraintsConfig, error) {
+	if reader == nil || reflect.ValueOf(reader).IsNil() {
+		return &api.PodNodeConstraintsConfig{}, nil
+	}
+
+	configBytes, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return nil, err
+	}
+	config := &api.PodNodeConstraintsConfig{}
+	err = configlatest.ReadYAML(configBytes, config)
+	if err != nil {
+		return nil, err
+	}
+	// No validation needed since config is just 2 bools
+	return config, nil
+}
+
+func (oo *podNodeConstraints) Admit(aa admission.Attributes) error {
+	switch {
+	case oo.config == nil,
+		aa.GetResource() != kapi.Resource("pods"),
+		aa.GetKind().Kind != "Pod",
+		aa.GetSubresource() != "",
+		aa.GetOperation() != admission.Create:
+		return nil
+	}
+	pod, ok := aa.GetObject().(*kapi.Pod)
+	if !ok {
+		return admission.NewForbidden(aa, fmt.Errorf("unexpected object: %#v", aa.GetObject()))
+	}
+	if len(pod.Spec.NodeName) > 0 && oo.config.ProhibitNodeName {
```

Comment:
@smarterclayton 

>To the first one I don't find it convincing.  Please describe a
scenario where anything that allows escape of a container is anything
less than access to other nodes.  The virus example is weak - I can
think of nothing that would actually stop reaching 100% saturation,
including this PR.

Regardless of whether we can think of one.  You are giving someone a way to get to every node.  Even if they have nothing but read access to the other containers.  That's an extra hole.


> The second argument is weak as well - a node that's almost full but
has low CPU activity.  So by adding their preferential location, they
are increasing utilization of a resource that would otherwise be
wasted.  What's the alternative?  They sit on another node and use the
exact same amount of CPU they would use before?  They just did us a
favor.

There are multiple ways this isn't doing a favor:

- Gives everyone on those nodes who have occasional usage a worse experience
- Messes up utilization data on those machines which might have fit into a lower utilization cost bucket
- Remove the possibility of giving preferential treatment to other pods by placing them on these sorts of nodes


Beyond those reasons this removes the ability of the scheduler and operators from controlling the systems and provides a users with an experience that isn't desired by operators or the platform providers.